listener/listener.go

```diff
@@ -294,59 +305,67 @@ func (l *Listener) Stream(ctx context.Context) error {
 
 	go l.SendPeriodicHeartbeats(ctx)
 
-	pool := &sync.Pool{
-		New: func() any {
+	eventPool := &sync.Pool{
+		New: func() interface{} {
 			return &publisher.Event{}
 		},
 	}
 
-	tx := NewWalTransaction(l.log, pool, l.monitor)
+	wg := &sync.WaitGroup{}
+	pipeline := NewPipeline(Config{ProcessErrors: false})
+	tx := NewWalTransaction(l.log, eventPool, l.monitor, l.cfg.Listener.Include.Tables, l.cfg.Listener.Exclude.Tables)
 
-	for {
-		if err := ctx.Err(); err != nil {
-			l.log.Warn("stream: context canceled", "err", err)
+	pipeline.Pipe(1, func(raw interface{}) interface{} {
+		if raw == nil {
 			return nil
 		}
+		msg := raw.(*pgx.ReplicationMessage)
+		wg.Add(1)
 
-		msg, err := l.replicator.WaitForReplicationMessage(ctx)
-		if err != nil {
-			return fmt.Errorf("wait for replication message: %w", err)
+		if err := l.parser.ParseWalMessage(msg.WalMessage.WalData, tx); err != nil {
+			l.monitor.IncProblematicEvents(problemKindParse)
+			return fmt.Errorf("parse: %w", err)
 		}
 
-		if msg == nil {
-			l.log.Debug("got empty message")
-			continue
-		}
+		if tx.CommitTime != nil {
+			events := tx.CreateEventsWithFilter(ctx)
+			eventBatch := &eventBatch{
+				events: events,
+				lsn:    msg.WalMessage.WalStart,
+			}
 
-		if err = l.processMessage(ctx, msg, tx); err != nil {
-			return fmt.Errorf("process message: %w", err)
+			l.log.Debug("WAL message has been processed", slog.Uint64("wal", eventBatch.lsn))
+			tx.Clear()
+			return eventBatch
 		}
 
 		l.processHeartBeat(msg)
-	}
-}
 
-func (l *Listener) processMessage(ctx context.Context, msg *pgx.ReplicationMessage, tx *WalTransaction) error {
-	if msg.WalMessage == nil {
-		l.log.Debug("empty wal-message")
 		return nil
-	}
-
-	l.log.Debug("WAL message has been received", slog.Uint64("wal", msg.WalMessage.WalStart))
+	}).Pipe(runtime.NumCPU(), func(raw interface{}) interface{} {
+		if raw == nil {
+			return nil
+		}
 
-	if err := l.parser.ParseWalMessage(msg.WalMessage.WalData, tx); err != nil {
-		l.monitor.IncProblematicEvents(problemKindParse)
-		return fmt.Errorf("parse: %w", err)
-	}
+		batch := raw.(*eventBatch)
+		results := make([]publisher.PublishResult, len(batch.events))
+		subjectNames := make([]string, len(batch.events))
 
-	if tx.CommitTime != nil {
-		for event := range tx.CreateEventsWithFilter(ctx, l.cfg.Listener.Filter.Tables) {
-			subjectName := event.SubjectName(l.cfg)
+		// loop over once triggering the publish of all the events
+		for i, event := range batch.events {
+			subjectNames[i] = event.SubjectName(l.cfg)
+			results[i] = l.publisher.Publish(ctx, subjectNames[i], event)
+		}
 
-			if err := l.publisher.Publish(ctx, subjectName, event); err != nil {
+		// loop over a second time awaiting the result of each publish
+		for i, result := range results {
+			_, err := result.Get(ctx)
+			if err != nil {
 				l.monitor.IncProblematicEvents(problemKindPublish)
 				return fmt.Errorf("publish: %w", err)
```

Comment:
This seems kinda problematic, you'll increment a failure event for the first failure but then return out of this whole process. Leaving all the other `results` just hanging. Potentially missing a lot of problems and having hanging background work.